bftengine/src/bftengine/messages/MessageBase.cpp

```diff
@@ -235,5 +289,30 @@ MessageBase *MessageBase::deserializeMsg(char *&buf, size_t bufLen, size_t &actu
   return msg;
 }
 
+// Start methods for diagnostics server:
+std::string MessageBase::getNumBuffsAllocatedForExtrnIncomingMsgs() {
+  return (std::string(" Num buffer allocations for external incoming messages: " +
+                      std::to_string(MessageBase::numIncomingExtrnMsgsBufAllocs)));
+}
+std::string MessageBase::getNumBuffsFreedForExtrnIncomingMsgs() {
+  return (std::string(" Num buffer frees for external incoming messages: " +
+                      std::to_string(MessageBase::numIncomingExtrnMsgsBufFrees)));
+}
+std::string MessageBase::getNumAliveExtrnIncomingMsgsObjsPerType() {
+  std::ostringstream oss;
+  oss << " Alive extrnal incoming message objects per type: " << std::endl;
+  oss << " --------------------------------------- " << std::endl;
+
+  for (int i = 0; i < MsgCode::MaxMsgCodeVal; ++i) {
+    if (IncomingExtrnMsgReceivedAtLeastOnceFlags.test(i)) {
```

Comment:
How can you lock IncomingExtrnMsgReceivedAtLeastOnceFlags in one place and not locking it here?
As far as I underatand IncomingExtrnMsgReceivedAtLeastOnceFlags update is rare, but can happen at any time (a new message which never arrived might arrive at any time).
if you dont lock also here, you may get a race condition for sure since you no idea how std::bitset is implmented and you might try to read from it at the moment it is updated by a writer.
in order to allow multiple readers and a single writer you need to use a shared mutex in both places where read/write.
lock_shared here
lock in the writer.
if there are no multiple readers (currently i guess there arent, but might be in the future?) you can just use a lock.

Reply:
I was aware of this issue, but since here it's stats report I didn't care much about making a "little wrong" report, if indeed a bit is written somewhere else when it is read here. However, your point regarding the STL implementation is correct, I would never imagine someone reallocating a const size bitset to another address but it isn't guaranteed so I'll solve it.